Mathlib/CategoryTheory/Products/Bifunctor.lean

```diff
@@ -57,3 +58,72 @@ theorem diagonal' (F : C × D ⥤ E) (X X' : C) (f : X ⟶ X') (Y Y' : D) (g : Y
 #align category_theory.bifunctor.diagonal' CategoryTheory.Bifunctor.diagonal'
 
 end CategoryTheory.Bifunctor
+
+namespace CategoryTheory.Functor
+
+/-- Opposite of a bifunctor.
+-/
+@[simps!]
+def biop (F : Cᵒᵖ ⥤ D ⥤ E) : C ⥤ Dᵒᵖ ⥤ Eᵒᵖ := F.rightOp ⋙ Functor.opHom _ _
+
+end CategoryTheory.Functor
+
+/- Utilities to apply functors into functor categories. -/
+namespace CategoryTheory.Functor
+
+-- We would like to keep these opaque in order to avoid leaking implementation details.
+
+/- Action of two-variable functors on objects. -/
+def obj2 (H : C ⥤ D ⥤ E) (A : C) (B : D) : E := (H.obj A).obj B
```

Comment:
This seems like a good idea to me, but it has to be an abbreviation, and I would suggest using small indices (type "\2" is VSCode):
```suggestion
def obj₂ (H : C ⥤ D ⥤ E) (A : C) (B : D) : E := (H.obj A).obj B
```

Reply:
I left `abbrev` for all expressions that are not constructed with a composition, but wouldn't it be better to hide the implementation details of `obj₂?`?

Reply:
I think we should keep `obj₂`  as an abbrev. For the other definitions, I think the best solution would be to have also ` map₂` and ` map₃` as abbreviations (maybe even with `@[reducible]`?), and to have `naturality_app` as a global simp lemma (I hope this would not break too many things...), so that `naturality₂` could be proved `by simp` (and thus become not essential, even though we may keep it as a nonsimp lemma). And, in order to get `naturality₃` essentially automatically `by simp`, it would be sufficient to introduce another simp lemma `naturality_app_app`.

Reply:
I thought `abbrev` was already `@[reducible]`? Maybe I'm misremembering. Regardless I've switched these to `abbrev`s. I don't think naturality is "directed" enough to be a good simp lemma; it seems likely we'll want to rewrite in different directions depending on the natural transformation

Reply:
Hm, if we mark `map₂, map₃` as `abbrev` then `reassoc` interacts very poorly with lemmas that mention `map₂`, `map₃` (eg `dinaturality_assoc`). Making them reducible makes sense if we expect that lemmas about `map₂, map₃` should be proved by reducing everything to `map` and using automation. But I'm not sure this is what we should design for? When working with bifunctors/profunctors we're likely going to want to think of `map₂` as an atom. I've reverted the change to `abbrev`s because I think it's more trouble than it's worth.

Reply:
In monoidal categories, `f ⊗ g` is thought as `(f ▷ _) ≫ (_ ◁ g)` since the refactor. Then, as a matter of consistency, I think that if it is possible not to use a `def`, and use `map₂` only as a reducible abbrev, this is the direction we should go.

Reply:
Do we have analogues of ▷ and ◁ for the action of a bifunctor? If we add those then I think we could break up dinaturality like `μ_natural_left`/`μ_natural_right`. As is simp plays extremely poorly with `map₂`

Reply:
Or I guess the analogue is just mapping with an identity argument. I'll try and see if dinaturality can be broken into two cases like I said

Reply:
I would say it is ` (H.map f).app D₁` and `(H.obj C₂).map g` which are in the definitions of `map₂`.